packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -235,128 +163,168 @@ fn read_file(path: &str) -> Result<String> {
 fn eval_file<'s>(scope: &mut v8::HandleScope<'s>, path: &str) -> Result<Value<'s>> {
     let source_code = read_file(path)?;
     let js_source_code = new_js_string(scope, &source_code);
-    let mut try_catch = v8::TryCatch::new(scope);
-    let script = v8::Script::compile(&mut try_catch, js_source_code, None).ok_or_else(|| {
-        let exception = try_catch.exception().unwrap();
-        let exception_string = exception
-            .to_string(&mut try_catch)
-            .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
-
-        Error::Eval(path.into(), format!("Compile error: {exception_string}"))
-    })?;
-
-    script.run(&mut try_catch).ok_or_else(|| {
-        let exception = try_catch.exception().unwrap();
-        let exception_string = exception
-            .to_string(&mut try_catch)
-            .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
-
-        Error::Eval(path.into(), format!("Execution error: {exception_string}"))
+    let mut try_catch_scope = v8::TryCatch::new(scope);
+    let script =
+        v8::Script::compile(&mut try_catch_scope, js_source_code, None).ok_or_else(|| {
+            let exception = read_exception(&mut try_catch_scope).unwrap();
+
+            Error::Eval(path.into(), format!("Compile error: {exception}"))
+        })?;
+
+    script.run(&mut try_catch_scope).ok_or_else(|| {
+        let exception = read_exception(&mut try_catch_scope).unwrap();
+
+        Error::Eval(path.into(), format!("Execution error: {exception}"))
     })
 }
 
-fn import_file<'s>(
-    scope: &mut v8::HandleScope<'s>,
-    context: v8::Local<'s, v8::Context>,
-    path: &str,
-    args: &[Value<'s>],
-) -> Result<Value<'s>> {
-    let v = eval_file(scope, path)?;
-    let f: Function<'_> = v.try_into().map_err(|err| {
-        Error::FileImport(
-            path.into(),
-            format!("Could not convert file into Function: {err}"),
-        )
-    })?;
+/// Cache module to not evaluate them twice
+struct ModuleMap {
+    path_to_module: HashMap<String, v8::Global<v8::Module>>,
+}
+
+impl ModuleMap {
+    fn new() -> Self {
+        Self {
+            path_to_module: HashMap::new(),
+        }
+    }
+
+    fn import_module<'s>(
+        scope: &mut v8::HandleScope<'s>,
+        path: &str,
+    ) -> Result<v8::Local<'s, v8::Module>> {
+        let module_map = scope
+            .get_slot::<Rc<RefCell<ModuleMap>>>()
+            .expect("ModuleMap is not present in isolate slots")
+            .clone();
+
+        if let Some(module) = module_map.borrow().path_to_module.get(path) {
```

Comment:
Can borrowing fail here?

Reply:
I don't think it can in practice, multiple threads can't call this at the same time.
We'd have to `borrow_mut` before to make it fail.